### PR TITLE
[Grapbolt]Add unified data structure

### PR DESCRIPTION
**Pull request by peizhou001**
## Description
Add unified data structure and subclass about node and edge.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/graphbolt/unified_data_struct.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+"""Unified data structure for input and ouput of all the stages in loading process."""
+
+from dataclasses import dataclass
+from typing import Dict, Tuple, Union
+
+import torch
+
+from .sampled_subgraph import SampledSubgraph
+
+__all__ = ["UnifiedDataStruct"]
```
Comment (frozenbugs) [inline python/dgl/graphbolt/unified_data_struct.py:10]: What about just DataBlock, LinkPredictionBlock, NodeClassificationBlock?

File: python/dgl/graphbolt/unified_data_struct.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+@dataclass
+class UnifiedDataStruct:
+    r"""A composite data class for data structure in the graphbolt. It is
+    designed to facilitate the exchange of data among different components
+    involved in processing data. The purpose of this class is to unify the
+    representation of input and output data across different stages, ensuring
+    consistency and ease of use throughout the loading process."""
+
+    seed_node: Union[torch.Tensor, Dict[str, torch.Tensor]]
```
Comment (frozenbugs) [inline python/dgl/graphbolt/unified_data_struct.py:21]: Do we have seed_node for link prediction?

File: python/dgl/graphbolt/unified_data_struct.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    - If `node_feature` is a dictionary: The keys should be node type and the
+      value should be corresponding node feature or embedding.
+    """
+    edge_feature: Union[torch.Tensor, Dict[Tuple(str, str, str), torch.Tensor]]
+    """A representation of edge feature.
+    - If `edge_feature` is a tensor: It indicates the graph is homogeneous.
+    - If `edge_feature` is a dictionary: The keys should be edge type and the
+      value should be corresponding edge feature or embedding.
+    """
+    input_nodes: Union[torch.Tensor, Dict[Tuple(str, str, str), torch.Tensor]]
```
Comment (frozenbugs) [inline python/dgl/graphbolt/unified_data_struct.py:45]: Isn't this the first layer of sampled_subgraphs?

File: python/dgl/graphbolt/node_unified_data_struct.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+from .unified_data_struct import UnifiedDataStruct
+
+
+@dataclass
+class NodeUnifiedDataStruct(UnifiedDataStruct):
+    r"""A subclass of 'UnifiedDataStruct', specialized for handling node level
+    tasks."""
+
+    label: Union[torch.Tensor, Dict[str, torch.Tensor]] = None
```
Comment (frozenbugs) [inline python/dgl/graphbolt/node_unified_data_struct.py:17]: What does each item in the Tensor corresponding to which node?

File: python/dgl/graphbolt/link_unified_data_struct.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    r"""A subclass of 'UnifiedDataStruct', specialized for handling edge level
+    tasks."""
+
+    node_pair: Union[
+        Tuple[torch.Tensor, torch.Tensor],
+        Dict[Tuple(str, str, str), Tuple[torch.Tensor, torch.Tensor]],
+    ]
+    """
+    Representation of seed node pairs utilized in link prediction tasks. Please
+    note that after the subgraph sampling process, the IDs contained within
```
Comment (frozenbugs) [inline python/dgl/graphbolt/link_unified_data_struct.py:24]: Why the id here will be compressed? but not the others?

File: python/dgl/graphbolt/unified_data_struct.py (hunk 3 above)
Comment (peizhou001) [reply python/dgl/graphbolt/unified_data_struct.py:40]: Yes, the outmost layer.

File: python/dgl/graphbolt/data_block.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    designed to facilitate the exchange of data among different components
+    involved in processing data. The purpose of this class is to unify the
+    representation of input and output data across different stages, ensuring
+    consistency and ease of use throughout the loading process."""
+
+    sampled_subgraphs: list[SampledSubgraph]
+    """
+    A list of 'SampledSubgraph's, each one corresponding to one layer,
+    representing a subset of a larger graph structure.
+    """
```
Comment (frozenbugs) [inline python/dgl/graphbolt/data_block.py:25]:
```suggestion
    """

```

File: python/dgl/graphbolt/data_block.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    node_feature: Union[torch.Tensor, Dict[str, torch.Tensor]]
+    """A representation of node feature.
+    - If `node_feature` is a tensor: It indicates the graph is homogeneous.
+    - If `node_feature` is a dictionary: The keys should be node type and the
+      value should be corresponding node feature or embedding.
+    """
```
Comment (frozenbugs) [inline python/dgl/graphbolt/data_block.py:31]:
```suggestion
    """

```

File: python/dgl/graphbolt/data_block.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    edge_feature: list[
+        Union[torch.Tensor, Dict[Tuple(str, str, str), torch.Tensor]]
+    ]
+    """A representation of edge feature corresponding to 'sampled_subgraphs'.
+    - If `edge_feature` is a tensor: It indicates the graph is homogeneous.
+    - If `edge_feature` is a dictionary: The keys should be edge type and the
+      value should be corresponding edge feature or embedding.
+    """
```
Comment (frozenbugs) [inline python/dgl/graphbolt/data_block.py:39]:
```suggestion
    """

```

File: python/dgl/graphbolt/link_prediction_block.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    tasks."""
+
+    node_pair: Union[
+        Tuple[torch.Tensor, torch.Tensor],
+        Dict[Tuple(str, str, str), Tuple[torch.Tensor, torch.Tensor]],
+    ]
+    """
+    Representation of seed node pairs utilized in link prediction tasks. Please
+    note that after the subgraph sampling process, the IDs contained within
+    could be compressed. Additionally, the head and tail sizes may differ when
```
Comment (frozenbugs) [inline python/dgl/graphbolt/link_prediction_block.py:24]: remove "compressed" related doc?

File: python/dgl/graphbolt/node_classification_block.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    r"""A subclass of 'UnifiedDataStruct', specialized for handling node level
+    tasks."""
+
+    seed_node: Union[torch.Tensor, Dict[str, torch.Tensor]]
+    """
+    Representation of seed nodes used for sampling in the graph.
+    - If `seed_node` is a tensor: It indicates the graph is homogeneous.
+    - If `seed_node` is a dictionary: The keys should be node type and the
+      value should be corresponding heterogeneous node ids.
+    """
```
Comment (frozenbugs) [inline python/dgl/graphbolt/node_classification_block.py:23]: add a new line as well.

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/graphbolt/link_prediction_block.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    """
+    Representation of seed node pairs utilized in link prediction tasks.
+    - If `node_pair` is a tuple: It indicates a homogeneous graph where each
+      tuple contains two tensors representing source-destination node pairs.
+    - If `node_pair` is a dictionary: The keys should be edge type, and the
+      value should be a tuple of tensors representing node pairs of the given
+      type.
+    """
+
+    label: Union[torch.Tensor, Dict[str, torch.Tensor]] = None
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/link_prediction_block.py:30]: why it's not `Tuple[str, str, str]` for edge type, consistent with `node_pair`

File: script/build_dgl.sh [hunk 12]
```diff
@@ -20,11 +20,12 @@ OPTIONS:
   -e           Extra arguments of cmake.
   -g           Restart CUDA build.
   -r           Remove all intermediate output.
+  -t           Type of the build: dev, test or release (default: dev).
```
Comment (Rhett-Ying) [inline script/build_dgl.sh:23]: something goes wrong in rebase/merge?

**Pull request merged**